README
This module is a fast XSLT library, based on the GNOME libxslt engine
that you can find at http://www.xmlsoft.org/XSLT/

Performance is currently about twice that of XML::Sablotron (based on 
XSLTMark tests converted to Perl).

The libxslt processor is also highly standards compliant
implementation of XSLT 1.0.

A NOTE ON INSTALLATION
----------------------

The library depends on libxslt. You have to have both the library and
its header files installed (on various Linux/UNIX distributions thay
are usually packaged as libxslt-dev or libxslt-devel). The module can
further make use of libexslt, which is distributed with libxslt, but
on some systems packaged separately. These libraries may have their
own dependencies (usually resolved automatically by the packaging
system).

Note that if you want any global data to be shared between the libxml2
libraries used by this module and XML-LibXML, e.g. the input callback
table, then each module must link dynamically against libxml2 so that
they both use the same shared library (this is in fact default on most
UNIX/POSIX platforms).

The Makefile.PL tries to determine the correct compiler and linker
flags for its library dependencies using pkg-config and
xslt-config. If this fails, you may override the values like this

  perl Makefile.PL INC="..." LIBS="..." 

where INC contains flags for the compiler (such as -I/some_path/include etc.) and LIBS contains linker flags (such as -L/some_path/lib -llibsomething ...).

WIN32 BUILDING ISSUES ---------------------

It is recommended that when building this module on Win32 you also
link dynamically against libxslt and libexslt, since linking
statically against these libraries but dynamically against libxml2 may
cause some "unresolved symbol" linker errors.

Dynamic linking, however, causes a name clash between the name of this
module's DLL (LibXSLT.dll) and that of libxslt (libxslt.dll) on
Win32's case-insensitive filesystem. So, if you link dynamically then
you'll probably have to build your libxslt DLL as, say,
libxslt_win32.dll and then make the appropriate changes to this
module's build process to look for that instead of libxslt.dll.

PATCHES AND DEVELOPER VERSION
=============================

As XML::LibXSLT is open source software help and patches are appreciated. If you
find a bug in the current release, make sure this bug still exists in the
developer version of XML::LibXSLT. This version can be downloaded from its
Mercurial repository. The repository can be cloned via:
    
    hg clone ssh://hg@bitbucket.org/shlomif/perl-xml-libxslt

Note this account does not allow direct commits.

Please consider the tests as correct. If any test fails it is most certainly
related to a bug.

COPYRIGHT AND LICENSE
---------------------

This is free software, you may use it and distribute it under the same
terms as Perl itself. Copyright 2001 AxKit.com Ltd.


Changes
Revision history for Perl extension XML::LibXSLT.

1.99    2020-01-16
    - Remove checks for libxslt versions
        - https://rt.cpan.org/Ticket/Display.html?id=131155
        - Thanks to SREZIC and others
        - These checks were much maligned.

1.98    2020-01-16
    - Replace xslt-config with pkg-config
        - The former is going away
        - https://github.com/shlomif/perl-XML-LibXSLT/pull/3
        - Thanks to Debian.

1.97    2020-01-14
    - Tests hot fix for https://rt.cpan.org/Public/Bug/Display.html?id=131457
        - Tests failed with XML::LibXML v 2.0202
        - Thanks to jplesnik from Red Hat.

1.96    2018-02-21
    - Apply patch from RT#83028
        - $Config{libs} is not needed
    - Move to git/GitHub:
        - https://github.com/shlomif/perl-XML-LibXSLT .
        - For Travis-CI/etc.

1.95    2016-07-30
    - Hopefully fix RT#116461
        - https://rt.cpan.org/Public/Bug/Display.html?id=116461
        - “Makefile.PL shouldn't run "pkg-config libexslt" if LIBS or INC were
        provided”
        - Thanks to NWELLNHOF for the report and a suggested fix.

1.94    2015-01-19
    - Fix builds on older versions of libxslt.
        - Breakage caused by the MaxVars change.
        - https://rt.cpan.org/Ticket/Display.html?id=101605
        - Thanks to Deyan Ginev for the report.

1.93    2015-01-17
    - Add the ->max_vars function to set the MaxVars.
        - https://rt.cpan.org/Ticket/Display.html?id=101413
        - Thanks to Thomas Krichel for the report/request.

1.92    2014-04-12
    - Add scripts/tag-release.pl .

1.91    2014-04-12
    - t/cpan-changes.t : minimum version of Test::CPAN::Changes.
        - This is to avoid test failures such as:
            - http://www.cpantesters.org/cpan/report/5af7d5eb-6bfa-1014-b84e-eb85912f2992

1.90    2014-04-10
    - Handle $Config{dlext} correctly on Windows.
        - https://rt.cpan.org/Ticket/Display.html?id=94516
        - Thanks to kmx for the report and fix.
    - Add t/pod.t .
        - Fix some POD problems.
    - Add t/cpan-changes.t and convert Changes to CPAN::Changes format.
    - Add t/style-trailing-space.t.
        - Remove trailing whitespace.

1.89    2014-03-05
    - Fix memory management of node-sets returned from extension functions.
      Without calling xsltExtensionInstructionResultRegister, node-sets
      could be freed too early, leading to strange errors or segfaults.
        - https://bugzilla.gnome.org/show_bug.cgi?id=562302
        - https://rt.cpan.org/Ticket/Display.html?id=93349
        - Thanks to Alexander Schepanovski for the report.
    - Document input_callbacks methods
    - Don't register libxml2 input callbacks twice
        - https://rt.cpan.org/Public/Bug/Display.html?id=60313
    - Fix memory leak with extension functions returning node lists
        - https://rt.cpan.org/Public/Bug/Display.html?id=55623
        - https://rt.cpan.org/Public/Bug/Display.html?id=76596
        - Thanks to Jozef Kutej and Slaven Rezic for the reports.
    - Thanks to Nick Wellnhofer for the pull requests and fixes.

1.88    2014-02-17
    - Fix t/xml-grammar-failures.t based on the xml:space commentary here:
        - https://mail.gnome.org/archives/xml/2014-February/msg00000.html
        - Caused some failures which were ultimately our fault.

1.87    2014-02-03
    - Decrease the minimum libxslt version again.
        - Setting it as 1.1.28 did not fix the failures, and now we have
        some more tests to test against it.
        - https://rt.cpan.org/Ticket/Display.html?id=92659
        - Thanks to Slaven Rezic for the report.

1.86    2014-02-02
    - Add t/xml-grammar-failures.t to catch test failures in dependent modules.
        - Add some data files for it under t/data/*.
        - This fails on FreeBSD 9.2 currently.

1.85    2014-02-01
    - Add "use strict;" and "use warnings;" globally (CPANTS).
    - Add LICENSE to the MANIFEST.

1.84    2013-12-26
    - Hopefully fix a segmentation fault on Win32 introduced in the last ver.
        - https://rt.cpan.org/Ticket/Display.html?id=91652
        - Thanks to Alexandr Ciornii for the report, and to Nikolay Sivov
        for providing help.

1.83    2013-12-22
    - Get rid of compiler warnings when compiling using CLANG.
        - https://rt.cpan.org/Public/Bug/Display.html?id=91118
        - Thanks to Daniel Macks for the report.

1.82    2013-11-06
    - Makefile.PL: Report LibXSLT mismatch correctly.
        - https://rt.cpan.org/Ticket/Display.html?id=90085
        - Thanks to David Schmidt for the report, and to joel and brother
        from the channel #perl-help for analysis.

1.81    2013-07-05
    - Fix a typo reported by dsteinbrunner:
        - https://rt.cpan.org/Ticket/Display.html?id=86689

1.80    2013-01-23
    - Bump the minimum version of libxslt2 to 1.1.28.
        - due to failures such as http://www.cpantesters.org/cpan/report/6f5029ee-639a-11e2-8465-2a97270fb964

1.79    2012-11-26
    - Die when libxslt is 1.1.27.
        - Due to this bug - https://bugzilla.gnome.org/show_bug.cgi?id=684564

1.78    2012-09-06
    - Add output_method method, update documentation of media_type
    (dorian@cpan).
        - Thanks, Dorian!

1.77    2012-02-25
    - Apply a patch from Michael Ludwig clarifying the quoting of string
    parameters to the XSLT transformation.
        - Thanks, Michael!

1.76    2012-01-12
    - Made LibXSLT.xs compile with C compilers that don't accept declarations
    inside blocks:
        - https://rt.cpan.org/Public/Bug/Display.html?id=73905
        - These include Microsoft Visual C++ (in C-mode).

1.75    2011-10-28
    - Convert the following test scripts from Test.pm to Test::More :
        - t/03input.t
    - Fix for https://rt.cpan.org/Public/Bug/Display.html?id=71345 by Tim
    Brody:
        - «segfault in free()»
        - Thanks, Tim!

1.74    2011-10-25
    - Convert the following test scripts from Test.pm to Test::More :
        - t/01basic.t
        - t/02dromeds.t

Makefile.PL
# -------------------------------------------------------------------------- #
# $Id$
# -------------------------------------------------------------------------- #
# Makefile.PL for XML::LibXML.
# This file is required to generate a localized Makefile
# -------------------------------------------------------------------------- #

use strict;
use warnings;

use vars qw/$DEVNULL $is_Win32/;

BEGIN {
  if ($] < 5.008_001) {
    warn "\nSorry, at least Perl 5.8.1 is required for this module!\n\n";
    exit;
  }
}

use ExtUtils::MakeMaker;
use Config;


$|=0;

my %config;

while($_ = shift) {
    my ($key, $val) = split(/=/, $_, 2);
    $config{$key} = $val;
}

my $DEBUG = delete $config{DEBUG};

if ( $] < 5.008 or $config{NO_THREADS} ) {
  warn "disabling XML::LibXML support for Perl threads\n";
  $config{DEFINE} .= " -DNO_XML_LIBXML_THREADS";
}
delete $config{NO_THREADS};

# We use it several times later. See:
# https://rt.cpan.org/Public/Bug/Display.html?id=116461
my $HAVE_USER_DEFINED = (defined($config{LIBS}) or defined($config{INC}) );

unless ( $::is_Win32 ) { # cannot get config in W32
    my $xsltcfg = "xslt-config";
    my $libprefix = $ENV{XSLTPREFIX} || $config{XSLTPREFIX};

    delete $config{XSLTPREFIX}; # delete if exists, otherwise MakeMaker gets confused

    if ( defined $libprefix ) {
        $xsltcfg = $libprefix . '/bin/' . $xsltcfg;
    }

    # if a user defined INC and LIBS on the command line we must not
    # override them
    if ( ! $HAVE_USER_DEFINED ) {
        # get libs and inc from gnome-config
        eval {
            print "running xslt-config... ";
            my $ver = backtick("$xsltcfg --version");
            my ($major, $minor, $point) = $ver =~ /(\d+)\.(\d+)\.(\d+)/g;
            if (not
                (
                    ($major > 1)
                        or
                    (($major == 1) && ($minor > 1))
                        or
                    (($major == 1) && ($minor == 1) && ($point >= 18))
                )
            )
            {
                die +{ type => "ver", msg => <<'EOF'};
libxslt versions before 1.1.18 are buggy. Please install the latest version
EOF
            }
            elsif ($major == 1 and $minor == 1 and $point == 25)
            {
                die +{ type => "ver", msg => <<'EOF'};
libxslt-1.1.25 contains a deadlock that breaks the tests, and is not supported.
See: https://rt.cpan.org/Ticket/Display.html?id=50487 .
EOF
            }
            elsif ($major == 1 and $minor == 1 and $point == 27)
            {
                die +{ type => "ver", msg => <<'EOF'};
libxslt-1.1.27 does not handle namespaces well, and is not supported.
See: https://bugzilla.gnome.org/show_bug.cgi?id=684564 .
EOF
            }

            $config{LIBS} ||= backtick("$xsltcfg --libs");
            $config{INC} ||= backtick("$xsltcfg --cflags");
            print "ok\n";
        };
        my $Err = $@;
        if ($Err) {
            print "failed\n";
            if ((ref($Err) eq 'HASH') && (($Err->{type} || '') eq 'ver')) {
                print {*STDERR} ($Err->{msg}, "\n");
                exit 0; # 0 recommended by http://cpantest.grango.org (Notes for CPAN Authors)
            }
            warn "*** ", $@ if $DEBUG;
            warn "using fallback values for LIBS and INC\n";
            # backtick fails if gnome-config didn't exist...
            $config{LIBS} = '-L/usr/local/lib -L/usr/lib -lxslt -lxml2 -lz -lm';
            $config{INC} = '-I/usr/local/include -I/usr/include';

            print <<OPT;
options:
  LIBS='$config{LIBS}'
  INC='$config{INC}'
If this is wrong, Re-run as:
  \$ $^X Makefile.PL LIBS='-L/path/to/lib' INC='-I/path/to/include'

OPT
        }
    }
}

if ($config{LIBS} !~ /\-l(lib)?xslt\b/) {
    # in this case we are not able to run xml2-config. therefore we need to
    # expand the libz as well.
  if ($::is_Win32) {
    if( $ENV{ACTIVEPERL_MINGW} ) {
      $config{LIBS} .= ' -llibxslt.lib -llibxml2.lib';
    } else {
      $config{LIBS} .= ' -llibxslt -llibxml2';
    }
  } else {
    $config{LIBS} .= ' -lxml2 -lz';
  }
}

if ($config{LIBS} !~ /\-lz(lib)?\b/ and !($::is_Win32 && $config{LIBS} !~ /\-lzlib\b/)) {
    # note if libxml2 has not -lz within its cflags, we should not use
    # it! We should trust libxml2 and assume libz is not available on the
    # current system (this is ofcourse not true with win32 systems.
    # $config{LIBS} .= $::is_Win32 ? ' -lzlib' :' -lz';
    if ( $config{DEBUG} ) {
        warn "zlib was not configured\n";
        warn "set zlib\n" if $::is_Win32;
    }
  if ($::is_Win32) {
    if( $ENV{ACTIVEPERL_MINGW} ) {
      $config{LIBS} .= '';
    } else {
      $config{LIBS} .= ' -lzlib';
    }
  } else {
    $config{LIBS} .= ' -lz';
  }
}

if ($config{LIBS} !~ /\-lm\b/) {
    # math support is important, but is not available separately in W32
    $config{LIBS} .= $::is_Win32 ? '' :' -lm';
}

if (!have_library($::is_Win32 ? "libxslt" : "xslt")) {
    print STDERR <<DEATH;
libxslt not found
Try setting LIBS and INC values on the command line
Or get libxslt and libxml2 from
  http://www.libxml.org/
If you install via RPMs, make sure you also install the -devel
RPMs, as this is where the headers (.h files) are.
DEATH
exit 0; # 0 recommended by http://cpantest.grango.org (Notes for CPAN Authors)
}


if (have_library($::is_Win32 ? "libexslt" : "exslt")) {
    if (! $HAVE_USER_DEFINED) {
        my $exslt_defaults = $::is_Win32 ?
        ($ENV{ACTIVEPERL_MINGW} ? q/-llibexslt.lib/ : q/-llibexslt/) :
        q/-lexslt/; # -lgcrypt -lgpg-error/;
        my $exsltcfg = 'pkg-config libexslt';
        my ($exslt_libs,$exslt_inc);
        eval {
            print "running $exsltcfg... ";
            $exslt_libs = backtick("$exsltcfg --libs");
            $exslt_inc = backtick("$exsltcfg --cflags");
            $exslt_libs =~ s/-l(xml2|xslt|z|m)\s+//g;
            print "ok\n";
        };
        if ($@) {
            print "failed\n";
            warn "*** ", $@ if $DEBUG;
            warn "using fallback values for LIBS and INC\n";
            # backtick fails if gnome-config didn't exist...
            $exslt_libs  = $exslt_defaults;
            $exslt_inc   = '';
        }
        $config{LIBS} .= ' '.$exslt_libs;
        $config{INC} .= ' '.$exslt_inc;
    }
    $config{DEFINE} .= " -DHAVE_EXSLT"
}

if ($DEBUG) {
  print "LIBS: $config{LIBS}\n";
  print "INC: $config{INC}\n";
}

my $ldflags = delete $config{LDFLAGS};
if ($ldflags) {
    $config{dynamic_lib} = { OTHERLDFLAGS => " $ldflags " };
}

# Avoid possible shared library name conflict. On Win32 systems
# the name of system DLL libxlst.dll clashes with module's LibXSLT.dll.
# To handle this we are gonna rename module's DLL to LibXSLT.xs.dll.
if ($::is_Win32)
{
    if ($] eq '5.010000')
    {
        print STDERR <<'BUG';
******************************** !!!WARNING!!! ********************************
According to the version string you are running a win32 perl 5.10.0 that is
known to be suffering from dynaloader bug that prevents using non-default
module's DLL name. Therefore we have to use LibXSLT.dll that will clash with the
standard name of xslt library DLL. Please consider renaming your xslt library
to e.g. libxslt_win32.dll and then make the appropriate changes to this module's
build process to look for that instead of standard libxslt.dll. If you do not
make the proposed changes you will got during the tests many error popups with
slightly confusing message about "xsltApplyOneTemplate" function.
This bug is not present in perl 5.8.* and will be fixed in 5.10.1.
*******************************************************************************
BUG
    }
    else
    {
        # Fix for RT #94516 :
        # https://rt.cpan.org/Ticket/Display.html?id=94516
        $config{DLEXT} = 'xs.'.$Config{dlext};
    }
}

WriteMakefile(
    'NAME'	=> 'XML::LibXSLT',
    'VERSION_FROM' => 'LibXSLT.pm', # finds $VERSION
    'AUTHOR'    => 'Matt Sergeant',
    'ABSTRACT'  => 'Interface to GNOME libxslt library',
    'LICENSE'   => 'perl_5',
    'PREREQ_PM' =>
    {
        'Encode' => 0,
        'XML::LibXML' => "1.70",
        'strict' => 0,
        'warnings' => 0,
    },
    'OBJECT'     => '$(O_FILES)',
    (($ExtUtils::MakeMaker::VERSION >= 6.48)
        ? (MIN_PERL_VERSION => '5.008001',)
        : ()
    ),
    'META_MERGE' => {
        'meta-spec' => {
            'version' => 2,
        },
        'resources' => {
            'repository' => {
                'type' => 'hg',
                'url'  => 'https://bitbucket.org/shlomif/perl-xml-libxslt',
                'web'  => 'https://bitbucket.org/shlomif/perl-xml-libxslt',
            },
        },
    },
    %config,
);

###################################################################
# Functions
#  - these should really be in MakeMaker... But &shrug;
###################################################################

use Config;
use Cwd;
use Symbol;
use File::Spec;


BEGIN {
    $::is_Win32 = ($^O =~ /Win32/);
    if ($::is_Win32) {
        $DEVNULL = 'DEVNULL';
    }
    else {
        $DEVNULL = eval { File::Spec->devnull };
        if ($@) { $DEVNULL = '/dev/null' }
    }
}

sub rm_f {
    my @files = @_;
    my @realfiles;
    foreach (@files) {
        push @realfiles, glob($_);
    }
    if (@realfiles) {
        chmod(0777, @realfiles);
        unlink(@realfiles);
    }
}

sub rm_fr {
    my @files = @_;
    my @realfiles;
    foreach (@files) {
        push @realfiles, glob($_);
    }
    foreach my $file (@realfiles) {
        if (-d $file) {
            # warn("$file is a directory\n");
            rm_fr("$file/*");
            rm_fr("$file/.exists");
            rmdir($file) || die "Couldn't remove $file: $!";
        }
        else {
            # warn("removing $file\n");
            chmod(0777, $file);
            unlink($file);
        }
    }
}

sub xsystem {
  my (@command)=@_;
    if ($DEBUG) {
        print "@command\n";
        if (system(@command) != 0) {
            die "system call to '@command' failed";
        }
        return 1;
    }
    open(OLDOUT, ">&STDOUT");
    open(OLDERR, ">&STDERR");
    open(STDOUT, ">$DEVNULL");
    open(STDERR, ">$DEVNULL");
    my $retval = system(@command);
    open(STDOUT, ">&OLDOUT");
    open(STDERR, ">&OLDERR");
    if ($retval != 0) {
        die "system call to '@command' failed";
    }
    return 1;
}

sub backtick {
    my $command = shift;
    if ($DEBUG) {
        print $command, "\n";
        my $results = `$command`;
        chomp $results;
        if ($? != 0) {
            die "backticks call to '$command' failed";
        }
        return $results;
    }
    open(OLDOUT, ">&STDOUT");
    open(OLDERR, ">&STDERR");
    open(STDOUT, ">$DEVNULL");
    open(STDERR, ">$DEVNULL");
    my $results = `$command`;
    my $retval = $?;
    open(STDOUT, ">&OLDOUT");
    open(STDERR, ">&OLDERR");
    if ($retval != 0) {
        die "backticks call to '$command' failed";
    }
    chomp $results;
    return $results;
}

sub try_link0 {
    my ($src, $opt) = @_;
    my $cfile = gensym();
    # local $config{LIBS};
    # $config{LIBS} .= $opt;
    unless (mkdir(".testlink", 0777)) {
        rm_fr(".testlink");
        mkdir(".testlink", 0777) || die "Cannot create .testlink dir: $!";
    }
    chdir(".testlink");
    open($cfile, ">Conftest.xs") || die "Cannot write to file Conftest.xs: $!";
print $cfile <<EOT;
#ifdef __cplusplus
extern "C" {
#endif
#include <EXTERN.h>
#include <perl.h>
#include <XSUB.h>
#ifdef __cplusplus
}
#endif

EOT
    print $cfile $src;
    print $cfile <<EOT;

MODULE = Conftest          PACKAGE = Conftest

PROTOTYPES: DISABLE

EOT
    close($cfile);
    open($cfile, ">Conftest.pm") || die "Cannot write to file Conftest.pm: $!";
    print $cfile <<'EOT';
package Conftest;
$VERSION = 1.0;
require DynaLoader;
@ISA = ('DynaLoader');
bootstrap Conftest $VERSION;
1;
EOT
    close($cfile);
    open($cfile, ">Makefile.PL") || die "Cannot write to file Makefile.PL: $!";
    print $cfile <<'EOT';
use ExtUtils::MakeMaker;
my %config;
while($_ = shift @ARGV) {
    my ($k, $v) = split /=/, $_, 2;
    warn("$k = $v\n");
    $config{$k} = $v;
}
WriteMakefile(NAME => "Conftest", VERSION_FROM => "Conftest.pm", %config);
EOT
    close($cfile);
    open($cfile, ">test.pl") || die "Cannot write to file test.pl: $!";
    print $cfile <<EOT;
use Test; BEGIN { plan tests => 1; } END { ok(\$loaded) }
use Conftest; \$loaded++;
EOT
    close($cfile);
    xsystem($^X,'Makefile.PL',map "$_=$config{$_}", keys %config);
    xsystem($Config{make},
	    ($config{MAKEAPERL} ? qw(-f Makefile.aperl FIRST_MAKEFILE=Makefile.aperl) : ()),
	    'test'); #,"OTHERLDFLAGS=".$opt);
}

sub try_link {
    my $start_dir = cwd();
    my $result = eval {
        try_link0(@_);
    };
    warn $@ if $DEBUG && $@;
    chdir($start_dir);
    rm_fr(".testlink");
    return $result;
}

sub have_library {
    my ($lib, $func) = (@_, "blank");
    printf("checking for %s() in -l%s... ", $func, $lib) if $func ne "blank";
    printf("looking for -l%s... ", $lib) if $func eq "blank";

    my $result;
    if ($func) {
        my $libs = $::is_Win32 ? " $lib.lib  " : "-l$lib";
        if ($::is_Win32) {
            $result = try_link(<<"SRC", $libs);
#include <windows.h>
#include <winsock.h>
blank() { return 0; }
int t() { ${func}(); return 0; }
SRC
            unless ($result) {
                $result = try_link(<<"SRC", $libs);
#include <windows.h>
#include <winsock.h>
blank() { return 0; }
int t() { void ((*p)()); p = (void ((*)()))${func}; return 0; }
SRC
            }
        }
        else {

            $result = try_link(<<"SRC", $libs);
blank() { return 0; }
int t() { ${func}(); return 0; }
SRC
        }
    }

    unless ($result) {
        print "no\n";
        return 0;
    }

    if ($func ne "main") {
        $config{DEFINE} .= uc(" -Dhave_$func");
    }

    print "yes\n";
    return 1;
}

sub MY::postamble {
  return <<'MAKE_FRAG';

runtest: pure_all
	perl -MFile::Spec -MTest::Run::CmdLine::Iface -e \
        "local @INC = @INC;	unshift @INC, map { File::Spec->rel2abs(\$$_) } ('$(INST_LIB)', '$(INST_ARCHLIB)'); Test::Run::CmdLine::Iface->new({test_files => [glob(q{t/*.t})]})->run();"

distruntest: distdir
	cd $(DISTVNAME) && $(ABSPERLRUN) Makefile.PL
	cd $(DISTVNAME) && $(MAKE) $(PASTHRU)
	cd $(DISTVNAME) && $(MAKE) runtest $(PASTHRU)

MAKE_FRAG
}

